# [LAB-03] 지도학습-예측모델ㅣ01-프로젝트 개요(실습코드).ipynb

# 1-2. 데이터 소개 / 품질 점검 / 기술 통계량 

주택가격 예측 분석 - PBT 정리
1. 분석 배경 및 목적 (1)
□ 주택 시장 분석의 의미 (주영아, 2025)

🏠 부동산

움직여 옮길 수 없는 유형의 자산, 토지와 그 정착물 (Ling & Archer, 2022)
경제 주체(개인 및 법인)의 보유 자산 유형 중 주식 및 채권과 대체성을 가지는 투자자산
포트폴리오 구성에서 핵심 위치를 차지
주로 건물의 형태로 대표

🏡 주택

누구에게나 필요한 주거 공간이자 삶의 터전 (Muth & Goodman, 2001)
개인의 삶의 질 결정, 소유 재화 중 가장 고가의 자산
경제적, 사회적으로 다양한 가치를 지님
소비재인 동시에 투자재
다양한 요인에 영향을 받는 복합적 특성을 지님 (Cho, 1996)
키워드: 개인 선호 수요와 공급 거시 경제 정책
□ 주택시장에 대한 분석과 예측 (King, 2015)
사회 연구의 주된 관심사 중 하나
경제적 예측의 실용적 적용: 주거와 투자 의사결정을 지원하고 금융기관의 위험을 관리
정치·사회적으로 시장 안정화와 사회 불평등 완화를 도모: 대출과 세금 등 정부의 각종 정책 수립 지원, 그 효과를 확인
지역사회 발전: 지역사회를 이해하고 도시 계획 방향을 설정하여 지속 가능한 발전을 도움

출처: 연세대학교 공학대학원 프롭테크학과 - 부동산산과

2. 분석 배경 및 목적 (2)
□ 주택 가격 추정의 사회적 필요성

◎ 주택 가격 추정은 재산세 과표와 담보 가치 산정에 실제로 쓰이는 모형

주체	의사결정	필요한 정보
가계	어디에 집을 살 것인가	지역별 가격 수준과 그 이유
지방정부	재산세 과표 산정	대량평가(mass appraisal) 모형
금융기관	담보 가치 평가	자동평가모형(AVM)
□ 정확도만큼 설명 가능성이 중요
세금 고지서에 "AI 판단"이라 쓸 수 없다. 과표와 담보는 근거를 문서로 제시해야 하므로, 계수로 설명되는 모형이 요구됨
감정평가사가 집집마다 방문하는 것은 비용상 불가능하기 때문에 위치·면적·연식으로 가격을 추정하는 모형을 주로 사용
□ 머신러닝(Machine Learning) 활용의 필요성

시장의 복잡성과 다양한 영향요인을 정량적으로 파악하고, 정교한 예측을 통해 효율적인 전략 수립과 합리적인 의사결정을 지원

3. 분석 배경 및 목적 (3) — 도메인지식 어필
□ 주택 경기를 이해하려는 시도 (주영아, 2025)
시기	내용
17세기 → 19세기	17세기 확률 이론의 발달과 함께 통계학 등장 → 18세기 구체화 → 19세기 연구에 본격 활용
1933 Hoyt / 1995 Quigley	초창기 경제·도시 연구 (주택 시장과 경제 모델의 통합적 접근)
1974 Rosen / 2003 Goodman & Thibodeau	질적 연구 & 데이터를 수집하고 분석하는 실증 연구의 발전 (헤도닉 가격 모형 정립)
2013 De Haan & Diewert	도시 연구가들은 다양한 시도를 통해 주택 시장의 흐름을 분석하고 예측하려 함
2016 Rossini ~ 2024 Moghaddam & Cao	컴퓨팅 및 데이터 과학을 필두로 한 정보 과학과 주택 관련 사회 과학 분야 > 양쪽에서 AI 기술을 활용한 주택 시장 연구 지속적으로 증가
2020 Jha et al.	현존하는 대부분의 알고리즘이 주택 시장 연구에 사용됨
2023 Zhou et al.	새로운 알고리즘이 등장할 때마다 이를 기존의 알고리즘 성능과 비교하는 연구 활발

부동산 가격 및 거래 데이터, 경제 및 금융 데이터, 주택 특성 및 입지의 지역 데이터, 수요와 공급, 사회 정책 관련 데이터

데이터 형태: 크로스 섹션(과거 헤도닉 등 단순 회귀), 시계열(거시경제 관련 트렌드 분석), 패널(데이터 가용성과 분석기법의 발전)

데이터 가용성과 분석 기법의 발전: 패널데이터 사용증가 - 시간적 변화와 개별적 이질성을 동시에 고려할 수 있어 다차원적 분석 가능

4. 분석 배경 및 목적 (4)
□ 선행연구 조사 - 계수 해석을 통한 주택 가격 설명과 공간 데이터의 활용 근거
◎ 헤도닉 가격이론 → 가격 결정요인의 계수 해석에 대한 이론적 근거 제공
◎ 공간 자기상관 연구 → 공간적 의존성을 고려한 좌표 활용의 이론적 근거 제공
연구	주요 내용	시사점
Rosen (1974)	여러 특성이 묶인 재화의 가격을 특성별 잠재가격의 합으로 분해하는 이론	회귀계수를 "방 하나의 값", "해안 접근성의 값"으로 해석하는 근거
Pace & Barry (1997)	이 데이터를 만들어 공개한 원본논문	가까운 구역끼리 집값이 닮는 현상이 공간적 자기상관을 반영한 모형 제시
Pace & Barry (1997), Sparse Spatial Autoregressions, Statistics and Probability Letters 33(3), 291–297
Rosen (1974), Hedonic Prices and Implicit Markets, Journal of Political Economy 82(1), 34–55

이론이 없으면 회귀계수는 그냥 숫자다. Rosen의 이론이 있어야 "소득 1만 달러당 집값이 얼마나 오른다"는 해석이 성립한다.

5. 분석 범위 · 단위 · 질문
□ 분석 범위

◎ 1990년 단일 시점 자료이므로 인과가 아니라 연관을 본다.

구분	내용
시간	1990년 4월 1일 기준 단일 시점 (시계열 아님)
공간	캘리포니아 주 (위도 32.54~41.95, 경도 -124.35~-114.31)
화폐 단위	1990년 달러 (물가 조정 안 됨)
□ 분석 질문 (가설 수립)
번호	질문
Q1	집값을 좌우하는 요인은 무엇인가
Q2	그 요인은 얼마나 강하게 작용하는가
Q3	처음 보는 구역의 집값을 얼마나 정확히 맞히는가
Q4	이 모형을 실제 업무에 쓸 수 있는가

여기 나오는 50만 달러는 1990년 달러다. 오늘의 부동산 시세와 직접 비교해서는 안 된다.

## #01. 준비작업 

### 1. 라이브러리 참조 

In [27]:
from jussam import load_data
from helpers import *
from pandas import DataFrame
import numpy as np

### 2. 데이터 불러오기 

In [28]:
origin = load_data("california_housing")
origin.head()

📚 1990년 미국 인구조사(Census)를 기반으로 캘리포니아 각 지역의 인구·주택·소득 특성과 주택 중위가격을 담아, 회귀 분석과 머신러닝 모델 실습에 널리 사용되는 대표적인 주택 가격 데이터 (출처: https://www.kaggle.com/datasets/camnugent/california-housing-prices)

    field               description
--  ------------------  --------------
 0  longitude           경도
 1  latitude            위도
 2  housing_median_age  주택 중위 연령
 3  total_rooms         총 방 수
 4  total_bedrooms      총 침실 수
 5  population          인구
 6  households          가구 수
 7  median_income       중위 소득
 8  median_house_value  주택 중위 가격



,longitude,latitude,housing_median_age,total_rooms,total_bedrooms,population,households,median_income,median_house_value,ocean_proximity
0,-122.230,37.880,41,880,129.000,322,126,8.325,452600,NEAR BAY
1,-122.220,37.860,21,7099,1106.000,2401,1138,8.301,358500,NEAR BAY
2,-122.240,37.850,52,1467,190.000,496,177,7.257,352100,NEAR BAY
3,-122.250,37.850,52,1274,235.000,558,219,5.643,341300,NEAR BAY
4,-122.250,37.850,52,1627,280.000,565,259,3.846,342200,NEAR BAY


## #02. 데이터 구조 점검 

### 1. 데이터 크기 

In [29]:
print(f"데이터 크기: {origin.shape}")

데이터 크기: (20640, 10)


### 2. 타입 점검 

In [30]:
origin.info()

<class 'pandas.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype   
---  ------              --------------  -----   
 0   longitude           20640 non-null  float64 
 1   latitude            20640 non-null  float64 
 2   housing_median_age  20640 non-null  int64   
 3   total_rooms         20640 non-null  int64   
 4   total_bedrooms      20433 non-null  float64 
 5   population          20640 non-null  int64   
 6   households          20640 non-null  int64   
 7   median_income       20640 non-null  float64 
 8   median_house_value  20640 non-null  int64   
 9   ocean_proximity     20640 non-null  category
dtypes: category(1), float64(4), int64(5)
memory usage: 1.4 MB


### 3. 중복값 점검 

In [31]:
df1 = my_qtcheck.check_duplicates(origin)

중복된 행의 수 : 0


### 4. 형식, 표기 단위 점검(1) - 연속형 변수 

In [32]:
# 값을 검사할 필드 목록 
fields = ['housing_median_age', 'total_rooms', 'total_bedrooms', 
          'population', 'households', 'median_income', 'median_house_value']    

minmax = [] # 최소값과 최대값을 저장할 리스트 

for field in fields: 
    min_value = df1[field].min()
    max_value = df1[field].max()
    minmax.append({"min":min_value, "max":max_value})   

# 최소값과 최대값을 데이터 프레임으로 변환하여 출력 
minmax_df = DataFrame(minmax, index=fields)
minmax_df 

,min,max
housing_median_age,1.000,52.000
total_rooms,2.000,39320.000
total_bedrooms,1.000,6445.000
population,3.000,35682.000
households,1.000,6082.000
median_income,0.500,15.000
median_house_value,14999.000,500001.000


### 4. 형식, 표기 단위 점검 (2) - 범주형 변수

In [33]:
df1['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64

6. 데이터 소개 (1)
□ 1990년 미국 인구조사 전수 자료
항목	내용
조사	1990 United States Census (10년 주기 전수조사)
지역	캘리포니아 주 전역
관측 단위	인구조사 블록그룹
관측 수	20,640개 구역, 변수 10개
최초 공개	Pace & Barry (1997), StatLib 저장소
캘리포니아 인구 커버리지: 98.9% (표본조사가 아니라 전수조사 → 대표성 측면에서 유리하게 작용 가능)
데이터상 인구 합계 29,421,840명 ÷ 1990년 공식 인구 29,760,021명
□ 데이터 출처
단계	주체	한 일
① 원자료	미국 인구조사국	1990년 인구주택 전수조사
② 학술	Pace & Barry (1997)	블록그룹 단위로 집계해 StatLib에 공개
③ 재배포	Luís Torgo (포르투대)	회귀 벤치마크 저장소에 수록
④ 교육용 개편	Aurélien Géron (Hands-On ML)	결측치를 임의 삽입하고 변수 1개(ocean_proximity)를 추가
⑤ 배포	Kaggle	camnugent/california-housing-prices
7. 데이터 소개 (2)
□ 데이터 구조 점검
크기	자료형	중복행	행·열 이름
20,640 × 10 (정상)	숫자형 9 + 문자형 1 (변환 완료 상태)	0건 (정상)	재배치·삭제 불필요 (정상)
구분	변수	의미	단위	척도	범위
종속변수	median_house_value	주택 가격 중위값	달러	연속형	14999 ~ 500001
위치 특성	longitude	구역 중심 경도	도(°)	연속형(좌표)	—
위치 특성	latitude	구역 중심 위도	도(°)	연속형(좌표)	—
위치 특성	ocean_proximity	해양 근접도	5개 범주	명목형	<1H OCEAN(9136건), INLAND(6551건), NEAR OCEAN(2658건), NEAR BAY(2290건), ISLAND(5건)
구조 특성	housing_median_age	주택 연식 중위값	년	연속형	1 ~ 52
구조 특성	total_rooms	구역 내 총 방수	개	연속형(총량)	2 ~ 39320
구조 특성	total_bedrooms	구역 내 총 침실 수	개	연속형(총량)	1 ~ 6445
인구·경제 특성	population	구역 인구	명	연속형(총량)	3 ~ 35682
인구·경제 특성	households	구역 가구 수	가구	연속형(총량)	1 ~ 6082
인구·경제 특성	median_income	가구 소득 중위값	만 달러 (0.5 = 5,000$)	연속형	0.5 ~ 15

위치 특성 범위 비고: 경도는 캘리포니아 밖으로 벗어나지 않고, 위도도 캘리포니아 밖으로 벗어나지 않음.

8. 데이터 소개 (3) — 형식·표기·단위 점검 / 한계 인지
□ 값의 범위와 표기가 물리적으로 가능한 형태인지 확인

상한 절단의 단서가 이 단계에서 드러난다.

연속형 변수 7개 — 최솟값·최댓값 범위 점검

변수	범위	판정
housing_median_age	1 ~ 52	절단 의심
total_rooms	2 ~ 39,320	범위 정상
total_bedrooms	1 ~ 6,445	범위 정상
population	3 ~ 35,682	극소 구역 존재
households	1 ~ 6,082	분모 주의
median_income	0.4999 ~ 15.0001	절단 의심
median_house_value	14,999 ~ 500,001	절단 의심

명목형 변수 ocean_proximity — 표기 점검 (모두 정상)

고유값 5개	앞뒤 공백	대소문자	표기 오류·중복
<1H OCEAN·INLAND·NEAR OCEAN·NEAR BAY·ISLAND	없음	전부 대문자	없음

## #03. 데이터 품질 점검 

### 1. 결측치 확인 

In [34]:
my_qtcheck.check_missing_values(df1)

,Missing Count,Missing Ratio (%)
longitude,0,0.000
latitude,0,0.000
housing_median_age,0,0.000
total_rooms,0,0.000
total_bedrooms,207,1.003
population,0,0.000
households,0,0.000
median_income,0,0.000
median_house_value,0,0.000
ocean_proximity,0,0.000


### 2. 결측치 제거 
결측치 5% 미만이므로 삭제하고 진행 

In [35]:
df2 = df1.dropna(subset=["total_bedrooms"])
my_qtcheck.check_missing_values(df2)

,Missing Count,Missing Ratio (%)
longitude,0,0.000
latitude,0,0.000
housing_median_age,0,0.000
total_rooms,0,0.000
total_bedrooms,0,0.000
population,0,0.000
households,0,0.000
median_income,0,0.000
median_house_value,0,0.000
ocean_proximity,0,0.000


### 3. 품질 점검 데이터셋 저장 

In [36]:
df2.to_excel("california_housing_qtcheck.xlsx", index=False)

9. 데이터 품질 점검 (1)
□ 결측치 점검과 처리

◎ 이 단계에서는 삭제만 한다.

대체는 정보 누수 때문에 분석용 데이터 전처리 단계로 미룬다.

결측 현황과 처리

변수	결측 건수	비율	처리
total_bedrooms	207	1.00%	삭제 → 비율이 1.00%로 5% 미만이므로 삭제해도 편향이 생기지 않음
나머지 9개 변수	0	0%	해당 없음
삭제 후 행 수			20,640 → 20,433

정보 누수 경고: 대체는 여기서 하지 않는다
전체 평균으로 채우면 평가용 데이터의 정보가 학습에 새어 들어간다. 대체가 필요하면 III장에서 학습·검증 분할 이후 학습용 통계량으로만 수행한다.

## #04. 기술 통계량 

### 1. 범주형 변수의 기술 통계량 

In [37]:
category_desc = my_qtcheck.categorical_summary(df2,
                save_path="california_housing_qtcheck.xlsx")
category_desc

컬럼 'ocean_proximity'의 value_counts():


,count
ocean_proximity,
<1H OCEAN,9034
INLAND,6496
ISLAND,5
NEAR BAY,2270
NEAR OCEAN,2628


,ocean_proximity
count,20433
unique,5
top,<1H OCEAN
freq,9034


### 2. 연속형 변수의 기술 통계량 

In [38]:
desc_df = my_qtcheck.numerical_summary(df2, 
        save_path="california_housing_numerical_summary.xlsx")
desc_df

,count,mean,std,min,25%,50%,75%,max,rel_diff,rdiff_flag,iqr,upper_bound,lower_bound,upper_outliers,upper_outliers_ratio,lower_outliers,lower_outliers_ratio,outliers,outliers_ratio,skew,skew_interpret,kurt,kurt_interpret,log_need
longitude,20433.000,-119.571,2.004,-124.350,-121.800,-118.490,-118.010,-114.310,-0.009,similar,3.790,-112.325,-127.485,0,0.000,0,0.000,0,0.000,-0.296,symmetric,-1.333,platykurtic,none
latitude,20433.000,35.633,2.136,32.540,33.930,34.260,37.720,41.950,0.040,similar,3.790,43.405,28.245,0,0.000,0,0.000,0,0.000,0.465,symmetric,-1.120,platykurtic,none
housing_median_age,20433.000,28.633,12.592,1.000,18.000,29.000,37.000,52.000,0.013,similar,19.000,65.500,-10.500,0,0.000,0,0.000,0,0.000,0.062,symmetric,-0.801,platykurtic,none
total_rooms,20433.000,2636.504,2185.270,2.000,1450.000,2127.000,3143.000,39320.000,0.240,diff,1693.000,5682.500,-1089.500,1290,0.063,0,0.000,1290,0.063,4.159,right tail,32.714,leptokurtic,log
total_bedrooms,20433.000,537.871,421.385,1.000,296.000,435.000,647.000,6445.000,0.236,diff,351.000,1173.500,-230.500,1271,0.062,0,0.000,1271,0.062,3.460,right tail,21.986,leptokurtic,log
population,20433.000,1424.947,1133.208,3.000,787.000,1166.000,1722.000,35682.000,0.222,diff,935.000,3124.500,-615.500,1190,0.058,0,0.000,1190,0.058,4.960,right tail,74.061,leptokurtic,log
households,20433.000,499.433,382.299,1.000,280.000,409.000,604.000,6082.000,0.221,diff,324.000,1090.000,-206.000,1210,0.059,0,0.000,1210,0.059,3.414,right tail,22.094,leptokurtic,log
median_income,20433.000,3.871,1.899,0.500,2.564,3.537,4.744,15.000,0.095,similar,2.180,8.014,-0.707,670,0.033,0,0.000,670,0.033,1.645,right tail,4.943,leptokurtic,log
median_house_value,20433.000,206864.413,115435.667,14999.000,119500.000,179700.000,264700.000,500001.000,0.151,diff,145200.000,482500.000,-98300.000,1064,0.052,0,0.000,1064,0.052,0.978,right tail,0.328,leptokurtic,log


In [39]:
desc_df = my_qtcheck.numerical_summary(df2)
desc_df[['mean', '50%', 'rel_diff', 'rdiff_flag', 'skew', 'outliers_ratio', 'log_need']]

,mean,50%,rel_diff,rdiff_flag,skew,outliers_ratio,log_need
longitude,-119.571,-118.490,-0.009,similar,-0.296,0.000,none
latitude,35.633,34.260,0.040,similar,0.465,0.000,none
housing_median_age,28.633,29.000,0.013,similar,0.062,0.000,none
total_rooms,2636.504,2127.000,0.240,diff,4.159,0.063,log
total_bedrooms,537.871,435.000,0.236,diff,3.460,0.062,log
population,1424.947,1166.000,0.222,diff,4.960,0.058,log
households,499.433,409.000,0.221,diff,3.414,0.059,log
median_income,3.871,3.537,0.095,similar,1.645,0.033,log
median_house_value,206864.413,179700.000,0.151,diff,0.978,0.052,log


In [40]:
# 판정 등급별 변수 개수
print("[ rdiff_flag 판정별 변수 개수 ]")
print(desc_df['rdiff_flag'].value_counts())

# 상대 거리가 유한한 숫자로 계산되지 않은 변수
inf_fields = desc_df[np.isinf(desc_df['rel_diff'])]
print(f"\n[ rel_diff 가 무한대인 변수: {len(inf_fields)}개 ]")
print(inf_fields[['mean', '50%', 'rel_diff', 'rdiff_flag']])

# 왜도 상위 5개
print("\n[ 왜도 상위 5개 ]")
print(desc_df['skew'].sort_values(ascending=False).head(5))

[ rdiff_flag 판정별 변수 개수 ]
rdiff_flag
diff       5
similar    4
Name: count, dtype: int64

[ rel_diff 가 무한대인 변수: 0개 ]
Empty DataFrame
Columns: [mean, 50%, rel_diff, rdiff_flag]
Index: []

[ 왜도 상위 5개 ]
population       4.960
total_rooms      4.159
total_bedrooms   3.460
households       3.414
median_income    1.645
Name: skew, dtype: float64


캘리포니아

rdiff_flag: diff 5개, similar 4개 → large_diff는 0개
inf 변수도 0개 (Empty DataFrame)
왜도 상위: population, total_rooms, total_bedrooms, households, median_income (구역 집계 변수들)

kc_house

rdiff_flag: similar 9개, large_diff 6개, diff 4개
inf 변수 4개 (waterfront, view, sqft_basement, yr_renovated)
왜도 상위: sqft_lot(13.06)이 압도적

이렇게 같은 판정 기준, 같은 코드를 두 데이터셋에 적용했을 때

캘리포니아는 가장 심해봤자 diff까지만 나왔는데
킹카운티는 large_diff가 6개나 나옴


□ California Housing 데이터 20,433행 기준 — 기술 통계
변수	count	mean	std	min	25%	50%	75%	max
longitude	20433	-119.571	2.004	-124.350	-121.800	-118.490	-118.010	-114.310
latitude	20433	35.633	2.136	32.540	33.930	34.260	37.720	41.950
housing_median_age	20433	28.633	12.592	1.000	18.000	29.000	37.000	52.000
total_rooms	20433	2636.504	2185.270	2.000	1450.000	2127.000	3143.000	39320.000
total_bedrooms	20433	537.871	421.385	1.000	296.000	435.000	647.000	6445.000
population	20433	1424.947	1133.208	3.000	787.000	1166.000	1722.000	35682.000
households	20433	499.433	382.299	1.000	280.000	409.000	604.000	6082.000
median_income	20433	3.871	1.899	0.500	2.564	3.537	4.744	15.000
median_house_value	20433	206864.413	115435.667	14999.000	119500.000	179700.000	264700.000	500001.000

데이터 중심 수준

변수	평균	중앙값	차이
median_house_value	206,864	179,700	+27,164 (오른쪽 꼬리)
median_income	3.871	3.537	+0.334 (오른쪽 꼬리)
housing_median_age	28.6	29.0	-0.4 (거의 대칭)
10. 기술 통계량 확인 (2)
□ 데이터의 퍼짐과 안정성
변수	왜도	해석
median_house_value	+0.978	우편향 (로그 변환 검토)
median_income	+1.645	강한 우편향 (로그 변환 검토)
housing_median_age	+0.062	대칭
total_rooms	+4.159	매우 강한 우편향 (로그 변환 검토)
population	+4.960	매우 강한 우편향 (로그 변환 검토)
□ 이상치 확인
변수	절단값	건수	비율
median_house_value	500,001$	958	4.7%
housing_median_age	52년	1,265	6.2%
median_income	15.0(만)	50	0.2%

모듈 실행 결과에 의한 기계적 제거 주의!
기계적으로 자르면 고가 구역이 통째로 사라진다. 값 절단을 이해하지 못하고 제거하면 비싼 동네를 다 지우는 셈이 되기 때문에 주의. 이는 분석 마지막에서 한계로 정확히 명시할 것.